perception/autoware_ground_segmentation/src/scan_ground_filter/grid.hpp

```diff
@@ -17,93 +17,481 @@
-                       std::atan2(mode_switch_radius_ + grid_size_m_, virtual_lidar_z_)) -
-                     universe_utils::normalizeRadian(mode_switch_angle_rad_);
-    inv_grid_size_rad_ = 1.0f / grid_size_rad_;
-    tan_grid_size_rad_ = std::tan(grid_size_rad_);
-    grid_id_offset_ = mode_switch_grid_id_ - mode_switch_angle_rad_ * inv_grid_size_rad_;
+    grid_dist_size_ = grid_dist_size;
+    grid_azimuth_size_ = grid_azimuth_size;
+
+    // set grid linearity switch radius
+    grid_linearity_switch_num_ = static_cast<int>(grid_linearity_switch_radius / grid_dist_size_);
+    grid_linearity_switch_radius_ = grid_linearity_switch_num_ * grid_dist_size_;
+
+    // calculate grid parameters
+    grid_dist_size_rad_ =
+      pseudoArcTan2(grid_linearity_switch_radius_ + grid_dist_size_, origin_z_) -
+      pseudoArcTan2(grid_linearity_switch_radius_, origin_z_);
+    grid_dist_size_inv_ = 1.0f / grid_dist_size_;
 
+    // generate grid geometry
+    setGridBoundaries();
+
+    // initialize and resize cells
+    cells_.clear();
+    cells_.resize(radial_idx_offsets_.back() + azimuth_grids_per_radial_.back());
+
+    // set cell geometry
+    setCellGeometry();
+
+    // set initialized flag
     is_initialized_ = true;
   }
 
-  float getGridSize(const float radius, const size_t grid_id) const
+  // method to add a point to the grid
+  void addPoint(const float x, const float y, const float z, const size_t point_idx)
   {
-    // check if initialized
-    if (!is_initialized_) {
-      throw std::runtime_error("ScanGroundGrid is not initialized.");
-    }
+    const float x_fixed = x - origin_x_;
+    const float y_fixed = y - origin_y_;
+    const float radius = std::sqrt(x_fixed * x_fixed + y_fixed * y_fixed);
+    const float azimuth = pseudoArcTan2(y_fixed, x_fixed);
 
-    float grid_size = grid_size_m_;
-    constexpr uint16_t back_steps_num = 1;
+    // calculate the grid id
+    const int grid_idx = getGridIdx(radius, azimuth);
 
-    if (radius > mode_switch_radius_ && grid_id > mode_switch_grid_id_ + back_steps_num) {
-      // equivalent to grid_size = (std::tan(gamma) - std::tan(gamma - grid_size_rad_)) *
-      // virtual_lidar_z_
-      // where gamma = normalizeRadian(std::atan2(radius, virtual_lidar_z_), 0.0f)
-      grid_size = radius - (radius - tan_grid_size_rad_ * virtual_lidar_z_) /
-                             (1 + radius * tan_grid_size_rad_ / virtual_lidar_z_);
+    // check if the point is within the grid
+    if (grid_idx < 0) {
+      return;
     }
-    return grid_size;
+    const size_t grid_idx_idx = static_cast<size_t>(grid_idx);
+
+    // add the point to the cell
+    cells_[grid_idx_idx].point_list_.emplace_back(Point{point_idx, radius, z});
   }
 
-  uint16_t getGridId(const float radius) const
+  size_t getGridSize() const { return cells_.size(); }
+
+  // method to get the cell
+  inline Cell & getCell(const int grid_idx) { return cells_[static_cast<size_t>(grid_idx)]; }
```

Comment:
Is it better to add explicit check size here?
```suggestion
  inline Cell & getCell(const int grid_idx) {
    const size_t idx = static_cast<size_t>(grid_idx);
    if (idx >= cells_.size()) {
        throw std::out_of_range("Invalid grid index");
    }
    return cells_[idx];
}
```